Bluetooth: Classic: AVRCP: enable full Br avrcp browsing commands

makeshi opened the pull request:
enable full Br avrcp browsing commands

makeshi committed:
Bluetooth: AVRCP: enable full browsing commands

This change implements full AVRCP Browsing support on both CT/TG,
and included all public APIs and callbacks, and add shell commands
for testing.

Signed-off-by: Make Shi <make.shi@nxp.com>

MarkWangChinese commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+ */
+typedef enum __packed {
+	BT_AVRCP_SCOPE_MEDIA_PLAYER_LIST = 0x00U,
+	BT_AVRCP_SCOPE_VFS               = 0x01,
```
Others have suffix, only this one doesn't have `U`.

makeshi replied:
removed

MarkWangChinese commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -3720,46 +3917,435 @@ int bt_avrcp_tg_send_subunit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid)
+	if (!IS_TG_ROLE_SUPPORTED()) {
+		LOG_ERR("Target role not supported");
+		return -ENOTSUP;
+	}
```
I see that the similar codes exist for many functions. Suggest to use small function to do it, then make the codes more simple.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+	uint8_t  status;       /**< bt_avrcp_status_t */
+	uint16_t uid_counter;  /**< UID counter */
+	uint16_t num_items;    /**< Number of items in this response */
+	uint8_t  items[];      /**< Sequence of items, each begins with bt_avrcp_item_hdr */
```
I think the more comments are needed to explain which structure will be used to parse the `items`.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2634,17 +2715,167 @@ static int avrcp_tg_handle_set_browsed_player_req(struct bt_avrcp *avrcp,
+	if (err < 0) {
+		LOG_ERR("Failed to send get folder items error response (err: %d)", err);
+		net_buf_unref(rsp_buf);
+	}
```
I saw there are too many similar code blocks with this one. 

If an error code `-EOPNOTSUPP` is returned here, the general reject is returned with error code `BT_AVRCP_STATUS_INTERNAL_ERROR` in caller `browsing_avrcp_recv()`.

makeshi replied:
yes, I think for this error cases can handled general reject, updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -3720,46 +3917,435 @@ int bt_avrcp_tg_send_subunit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid)
+		LOG_ERR("Invalid AVRCP target");
+		return -EINVAL;
+	}
+	if (!IS_TG_ROLE_SUPPORTED()) {
```
Empty line is missing.

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+typedef enum __packed {
+	BT_AVRCP_MEDIA_TYPE_AUDIO = 0x00,
+	BT_AVRCP_MEDIA_TYPE_VIDEO = 0x01,
+} bt_avrcp_media_type_t;
```
have removed all u

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+ * The @p name_attr field contains:
+ * - struct media_element_item_name (name_len + name[]).
+ * - Followed by struct media_element_item_attr.
+ */
```
There are two variable‑length arrays in name_attr, so I split the structure into two structs and placed each variable‑length array at the end.

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+	uint8_t  uid[8];       /**< 64-bit element UID. */
+	uint8_t  media_type;   /**< @ref bt_avrcp_media_type_t. */
+	uint16_t charset_id;   /**< Character set for the name. */
+	uint8_t  name_attr[];  /**< Flexible array for name and attributes. */
```
done

makeshi committed:
bluetooth: avrcp: align enums and response struct naming

Drop the 'U' suffix from enum hex literals keep style consistent
and rename `bt_avrcp_list_app_setting_attr_rsp` to the clearer.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -94,6 +94,9 @@ NET_BUF_POOL_FIXED_DEFINE(avctp_ctrl_tx_pool, CONFIG_BT_MAX_CONN,
 			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
 
+NET_BUF_POOL_DEFINE(avctp_browsing_rx_pool, BT_BUF_ACL_RX_COUNT,
+		    BT_BUF_ACL_SIZE(CONFIG_BT_BUF_ACL_RX_SIZE),
```
Have add configured for it, default is  minimum MTU of 335 for the AVCTP Browsing Channel

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -757,7 +926,84 @@ struct bt_avrcp_ct_cb {
 	 *             big-endian format.
 	 */
-	void (*browsed_player_rsp)(struct bt_avrcp_ct *ct, uint8_t tid, struct net_buf *buf);
+	void (*set_browsed_player)(struct bt_avrcp_ct *ct, uint8_t tid, struct net_buf *buf);
```
Done

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -757,7 +926,84 @@ struct bt_avrcp_ct_cb {
+	 *  @param buf The response buffer containing the Get Item Attributes response
+	 *             parameters as defined by AVRCP PDU GET_ITEM_ATTRIBUTES. The
+	 *             application should parse this payload per the AVRCP specification.
+	 *             Note that the data is encoded in big-endian format.
```
updated

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,195 @@ typedef enum __packed {
+	BT_AVRCP_SCOPE_MEDIA_PLAYER_LIST = 0x00U,
+	BT_AVRCP_SCOPE_VFS               = 0x01,
+	BT_AVRCP_SCOPE_SEARCH            = 0x02U,
+	BT_AVRCP_SCOPE_NOW_PLAYING       = 0x03U,
```
updated

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,224 @@ typedef enum __packed {
+struct media_element_item_attr {
+	uint8_t  num_attrs;           /**< Number of attributes */
+	struct bt_avrcp_media_attr attrs[]; /**< attribute tuples (id/charset/len/value) */
+} __packed;
```
Miss doxygen comments.

makeshi commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -513,6 +513,23 @@ config BT_AVRCP_BROWSING
+config BT_AVRCP_BROWSING_L2CAP_MTU
+	int "L2CAP MTU for AVRCP Browsing Channel"
+	default 339
+	depends on BT_AVRCP_BROWSING
```
fixed

makeshi committed:
Bluetooth: AVRCP: Add alloc_buf callback and browsing RX pool

Introduce an optional alloc_buf() callback into AVCTP so that each
session can decide which RX buffer pool to use. Provided a dedicated
RX pool for the AVRCP browsing channel.
Also provided BT_AVRCP_BROWSING_L2CAP_MTU for configure.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -331,6 +331,226 @@ typedef enum __packed {
+	BT_AVRCP_SCOPE_MEDIA_PLAYER_LIST = 0x00,
+	BT_AVRCP_SCOPE_VFS               = 0x01,
+	BT_AVRCP_SCOPE_SEARCH            = 0x02,
+	BT_AVRCP_SCOPE_NOW_PLAYING       = 0x03,
```
@kartben
Could you review this PR again? Thanks.  This PR is blocking the AVRCP PTS test PR [tests: bluetooth: tester: Add AVRCP and AVCTP PTS support by CanWang001 · Pull Request #99198 · zep…](https://github.com/zephyrproject-rtos/zephyr/pull/99198).

kartben reviewed (approved):
thanks for addressing the docs issues - +1 on that front
